### ISSUE TITLE
Bounty: Public database syncing system (150$)

### ISSUE THREAD

**Issue by ajayyy** [ajayyy/SponsorBlockServer#373]
There should be a system that allows anyone to be able to have a near-realtime (5 mins optimal, 20 mins at least) mirror at least the `sponsorTimes` table.

This was discussed in Discord([message](https://discord.com/channels/603643120093233162/603643205552177155/892332418231590932)) / Matrix([message](https://matrix.to/#/!CEXFpYSWnWxckVAWlu:ajay.app/$OT4k_fFnsd8mT5L0NDQ3DuYRfEB29nVfSdGXwg-fKLA?via=ajay.app&via=matrix.org&via=server.matrix4ulm.de)).

Possible options:
- Use a system similar to [MusicBrainz](https://musicbrainz.org/doc/Live_Data_Feed) where a Postgres replication feed gets mirrored in a read-only way for other users to download. There then should be a way for the server to automatically read from this replication feed and import it into their local database. This would have to be done in a secure way that only replicated the public tables.

- Reuse the existing [Webhook implementation](https://github.com/ajayyy/SponsorBlockServer/pull/118) and create a recieving end for this that can be configured to save this info to the database.

- Host the CSVs via IPFS and have a wrapper to automatically retrieve latest versions of this using IPFS. This should work as IPFS stores files as chunks, and will know which chunks and new and which are old. This also makes it so when someone is first downloading the entire db, everyone's mirror will act as a seeder, even if they are on a slightly outdated version of the file (since they have access to the specific chunks).

***

The solution should be able to be run easily when starting up the SponsorBlock server. Docker would probably be needed to ease deployment if it requires external dependencies outside of just node.

***

Bounty of $150 USD via PayPal to someone who solves this issue. To avoid duplicate work, make sure to comment on this issue with what you are doing.

**Comment by mchangrh** [comment]
I'm not sure how well useful IPFS chunking would be as a large amount of the CSV is modified in day-to-day operations, most notably votes. You can try this yourself by chunking and deduping different days from my historical archive

http://sb-archive.mchang.xyz/mirror/

**Comment by ajayyy** [comment]
Postgres approach is probably the best, but it does restrict all mirrors to have to use Postgres instead of Sqlite

**Comment by mchangrh** [comment]
braindump while I finish my readings

https://gist.github.com/mchangrh/d8e2837b372756c867429eb20e946798

**Comment by mchangrh** [comment]
I think I have a *very basic* implementation that would only require
- rsync
- cron

and high bandwidth RSync+nginx server(s) for hosting but that's quite obvious + I am willing to provide (offtopic)

https://gist.github.com/mchangrh/3c3a9e870dfc1b8aee574c791dab83ad

# tl;dr
## master (Ajay)
- base_backup every 12 hours
- ship WALs to storage once done ( can also ship via rclone to "cloud storage")
- prod storage w/ ssh/ webhook/ anything

## storage (single/ geodistributed)
- retain full history OR last base + WAL
- rsync, nginx
- (on prod) - generate archive (tar with base + all WAL)
- (on prod) - generate .db for sqlite3 (wal replays are too cumbersome on stock sqlite)
- (on prod) - prod mirrors

## active mirror
- (on prod) | cron - get latest WAL(s)
- any new WALs are automatically replayed

## mirror from scratch / back from the dead
- docker-compose pull etc...
- get latest archive OR base + WAL
- restore base, start in recovery to stream in WALs

**Comment by ajayyy** [comment]
> high bandwidth RSync+nginx server(s)

Is there a reason this would require more bandwith than now? As, kind of the point of this is to reduce bandwith used by many people downloading CSVs. I think the 16mb WALs should make it not take too much.

**Comment by mchangrh** [comment]
>Is there a reason this would require more bandwith than now? As, kind of the point of this is to reduce bandwith used by many people downloading CSVs. I think the 16mb WALs should make it not take too much.

the point would be to offload the bandwidth from the main server on to these mirrors, but they would probably be serving exports as well as these base_dumps and WALs, but if they are slower than the main server or just generally bad, they'll probably just fall back on the main server

**Comment by ajayyy** [comment]
I guess that part really depends on the amount of uptake on it past services like sb.ltn.fi

**Comment by mchangrh** [comment]
the main advantage is that the tier 1 mirrors can be dumb

small amount of storage (1-3GB) with minimal CPU usage and some RAM since all they have are
- nginx server
- rsync server
- nginx WH listener / sshd.

if the primary storage server isn't generating the archive, then tar which is minimal and db generation can be on a docker runner

**Comment by stranger-danger-zamu** [comment]
I found the addon today and it's great! 

You could also abuse HTTP range requests and host an index to the CSV lines based on time. So the program just asks the next CSV slice based on it's newest ID and the server just sends the correct range (and possibly a URL). Then the program just downloads just the new lines.

The beauty of this approach is that any plain old nginx instance could do this without any major changes. Additionally since we the server is serving as an index, then we can point to mirrors and round robin the load.

---

Real talk though. Compress those CSV files. Even gzip is good enough:

|original file| original size| gzip'd size| xz'd size|
|---|---|---|---|
|categoryVotes.csv      |3.4M|   1.2M|   996K|
|lockCategories.csv     |3.0M | 292K|  216K|
|sponsorTimes.csv|       635M|    272M|    208M|
|unlistedVideos.csv|     5.5M|  1.8M|  1.2M|
|userNames.csv|          5.1M|      2.6M|      2.4M|
|vipUsers.csv|           4.0K|        4.0K|        4.0K|
|warnings.csv|           84K|       32K|       28K|

GZIP and XZ are both installed on most systems and anyone downloading these files from the website should know how to deal with it. 

If they are being used programmatically for updating, there are ways of just opening the compressed file and reading it in all the popular languages.

**Comment by mchangrh** [comment]
would this work for modified lines? A lot of the delta is not just new segments being added but votes and views on existing segments too

**Comment by stranger-danger-zamu** [comment]
> would this work for modified lines? A lot of the delta is not just new segments being added but votes and views on existing segments too

I don't see why you couldn't just have the update logic just overwrite the record if it encounters on the same ID.

**Comment by mchangrh** [comment]
1/2 done - distribution and downloading of files

gist explaining flags and why lz4 was chosen (w/ benchmarks)
https://gist.github.com/mchangrh/3d4a967732f8994cfcfdc05a8e22dc4f

docker container for mirroring csv files
https://github.com/mchangrh/sb-mirror

aports merge for lz4 https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/26616 (if closed, will probably use debian as a base instead or build ourselves)
